chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,809 @@
     }
 
-    pub fn get_block_info(&mut self, block_hash: &B::Id) -> Result<BlockInfo<B::Id>, Error> {
-        let index = self.index.read().unwrap();
+    /// Prune a branch with the given tip id from the storage.
+    pub fn prune_branch(&self, tip_id: &[u8]) -> Result<(), Error> {
+        let tips = self.volatile.open_tree(tree::BRANCHES_TIPS)?;
 
-        let tx = self
-            .inner
-            .transaction()
-            .map_err(|err| Error::BackendError(Box::new(err)))?;
+        if !tips.contains_key(tip_id)? {
+            return Err(Error::BranchNotFound);
+        }
 
-        Self::get_block_info_internal(&tx, &index, block_hash)
-    }
+        let blocks = self.volatile.open_tree(tree::BLOCKS)?;
+        let info = self.volatile.open_tree(tree::INFO)?;
+        let chain_length_to_block_ids = self.volatile.open_tree(tree::CHAIN_LENGTH_INDEX)?;
 
-    pub fn get_block_infos_by_chain_length(
-        &mut self,
-        chain_length: u64,
-    ) -> Result<Vec<BlockInfo<B::Id>>, Error> {
-        self.do_by_chain_length(chain_length, Self::get_block_info_internal)
-    }
+        let result = (&blocks, &info, &chain_length_to_block_ids, &tips).transaction(
+            |(blocks, info, chain_length_to_block_ids, tips)| {
+                let mut result = RemoveTipResult::NextTip(Vec::from(tip_id));
 
-    fn get_block_info_internal(
-        connection: &Connection,
-        index: &ChainStorageIndex<B>,
-        block_hash: &B::Id,
-    ) -> Result<BlockInfo<B::Id>, Error> {
-        let row_id = index
-            .get_block_info(block_hash)
-            .ok_or(Error::BlockNotFound)?;
-
-        connection
-            .prepare_cached(
-                "select depth, parent, fast_distance, fast_hash from BlockInfo where rowid = ?",
-            )
-            .map_err(|err| Error::BackendError(Box::new(err)))?
-            .query_row(&[Value::Integer(*row_id as i64)], |row| {
-                let mut back_links = vec![BackLink {
-                    distance: 1,
-                    block_hash: blob_to_hash(row.get(1)?),
-                }];
-
-                let fast_distance: Option<i64> = row.get(2)?;
-                if let Some(fast_distance) = fast_distance {
-                    back_links.push(BackLink {
-                        distance: fast_distance as u64,
-                        block_hash: blob_to_hash(row.get(3)?),
-                    });
+                while let RemoveTipResult::NextTip(current_tip) = &result {
+                    result = match remove_tip_impl(
+                        blocks,
+                        info,
+                        chain_length_to_block_ids,
+                        tips,
+                        &self.permanent,
+                        current_tip,
+                        self.root_id.as_ref(),
+                        self.id_length,
+                    )? {
+                        Ok(res) => res,
+                        Err(err) => return Ok(Err(err)),
```

Comment:
To make this clear: this lets all previous removals in the transaction be committed, even though this one fails?
Then the method returns `Err` anyway?
Would it be better to roll back the whole transaction?